daffodil-core/src/main/scala/org/apache/daffodil/core/dpath/Expression.scala

```diff
@@ -2329,6 +2342,11 @@ abstract class FunctionCallBase(
   protected def checkArgArrayOrOptional(): Unit = {
     lazy val isArrayOrOptional = expressions.head match {
       case pe: PathExpression => pe.isPathToOneWholeArrayOrOptional
+      case ie: IfExpression => {
+        val tp = ie.thenPart
+        val ep = ie.elsePart
+        tp.isPathToOneWholeArrayOrOptional || ep.isPathToOneWholeArrayOrOptional
```

Comment:
Agreed, I would expect an `&&` here too.

Another thing maybe worth considering is that this doesn't allow nested ifs, since it only checks one level deep. So if the `then` or `else` branches contain another if-statement, even if those resolve to a path to a whole array/optional, it will fail.

Though, I'm not sure if that matters, and in fact, I'm wondering if we should just require that `fn:count()` accepts exactly a path--no conditionals, no other functions, etc. So we don't allow `fn:count(if(cond) then /path1 else /path2)` and that fails with an SDE like "fn:count argument must be a path". If you want conditional counts then you do it like this: `if (cond) then fn:count(/path1) else fn:count(/path2)`. It's a bit more verbose, but is more clear and avoids potential edge cases.

Reply:
I like the simplicity of allowing only a path as fn:count's argument.  I checked the DFDL specification and it was not very clear whether we must allow any DFDL expression to be given to fn:count as argument:

| Function | Meaning |
|--------|--------|
| fn:count($arg) | Returns the number of items in the value of $arg as an xs:integer. Returns 0 if $arg is the empty sequence. |

The following are Schema Definition Errors, regardless of whether they are detected in advance of processing or once processing begins:

  -   Expression value is not single node
    -  Most DFDL expression contexts require an expression to identify a single node, not an array (aka sequence of nodes). There are a few exceptions such as the fn:count(…) function, where the path expression must be to an array or optional element.
  -   Expression value is not array element or optional element.
    -  Some DFDL expression contexts require an array or an optional element.
    -  Example: The fn:count(...) function argument must be to an array or optional element. It is a Schema Definition Error if the argument expression is otherwise.

Mike, please advise us with your deeper knowledge of DPath and XPath.

Reply:
Ok. I think the DFDL spec is not clear on whether fn:count must take a path (not a DPath expression, a literally step/step/step thingy) or if it can be if/then/else so long as the then/else are both arrays or both optionals. 

I think this code doesn't work right for nested ifs. To get that to work this would have to compose like the way the type system composes. I.e., the 
```
tp.isPathToOneWholeArrayOrOptional || ep.isPathToOneWholeArrayOrOptional
```
(which should be && not ||) that would need to inductively be computed for the then-part and else part. 
Putting in that infrastructure, and then testing it sufficiently feels very not worth it. 

I think I'd rather keep it simpler and just say fn:count's argument must literally be a path (i.e., relative or absolute path steps). It's a fair interpretation of the DFDL spec to put this restriction in.